### PR TITLE
Fix several edge cases with context expression state

### LINKED ISSUE #7993: ES spec violation
## Bug Report

**Current Behavior**

The parser throws a SyntaxError

**Input Code**

This is a valid JavaScript code according to the latest ES standard:

```js
a=>{}
/\//
```

**Expected behavior/code**

No SyntaxError should be thrown

**Babel Configuration (.babelrc, package.json, cli command)**

Not sure, I'm using the [online babel parser](https://babeljs.io/)

**Possible Solution**

`’\_(ツ)_/‘`

**Additional context/Screenshots**

### LINKED ISSUE #8891: Parsing error when JSX argument contains object with `class` field
## Bug Report

**Input Code**

```js
<Select prop={{ class: 'test' }} />;
```

**Current Behavior**

With the latest version of `@babel/parser@7.1.3` it fails with the error

```
SyntaxError: /Users/serdiuk/coding/test/test.js: Unexpected token, expected "jsxTagEnd"
```

**Expected behavior/code**

With `@babel/parser@7.1.2` and bellow it works:

```
React.createElement(Select, {
  prop: {
    class: 'test'
  }
});
```

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
    "presets": ["@babel/preset-react"]
}

```

**Environment**
- Babel version(s): `@babel/core@7.1.2`
- Node/npm version: `v8.9.4`
- OS: OSX 10.12.6
- Monorepo: n/a
- How you are using Babel: cli, but also happens in `babel-loader`

**Possible Solution**

As a current workaround I can write the following:

```js
const prop = { class: 'test' };
<Select prop={prop} />;
```

This works with any version. But this is still an issue, because it fails to parse the code, that is supposed to be valid Javascript.

### PR DESCRIPTION
**Pull request by danez**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #8891 Fixes #7993
| Patch: Bug Fix?          | y
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

I started fixing #8891 and ended in an endless loop of breaking things with the fixes and having to fix that.

I took a lot of fixes from acorn, and reported a bunch of issues to acorn too.

### CODE REVIEW

File: packages/babel-parser/src/tokenizer/index.js [hunk 1]
```diff
@@ -1359,6 +1379,14 @@ export default class Tokenizer extends LocationParser {
     return !this.state.exprAllowed;
   }
 
+  inGeneratorContext() {
```
[the remark this answers is not in the document]
Comment (danez) [reply packages/babel-parser/src/tokenizer/index.js]: Yes works. Acorn does have both things too, but with the difference that they also allow running the tokenizer without the parser, that's why they need both `inGenerator` and `inGeneratorContext`.

File: packages/babel-parser/src/tokenizer/context.js [hunk 2]
```diff
@@ -46,30 +64,44 @@ tt.parenR.updateContext = tt.braceR.updateContext = function() {
...
+      allowed = true;
+    }
+  }
+  this.state.exprAllowed = allowed;
 
   if (prevType === tt._let || prevType === tt._const || prevType === tt._var) {
-    if (lineBreak.test(this.input.slice(this.state.end))) {
+    if (
+      lineBreak.test(
+        this.input.slice(this.state.end, this.lookahead(true).start),
```
[the remark this answers is not in the document]
Comment (danez) [reply packages/babel-parser/src/tokenizer/context.js]: I removed this and instead used the same fix that acorn did by calling `readRegex()` in `parseExprAtom()`

File: packages/babel-parser/src/plugins/flow.js [hunk 3]
```diff
@@ -2392,7 +2393,10 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       refNeedsArrowPos?: ?Pos,
     ): N.Expression {
       let jsxError = null;
-      if (tt.jsxTagStart && this.match(tt.jsxTagStart)) {
+      if (
+        this.hasPlugin("jsx") &&
```
Comment (danez) [inline packages/babel-parser/src/plugins/flow.js]: I had to replace this, as `tt.jsxTagStart` is always set in our tests, even if jsx plugin not loaded. Yeahh monkeypatching :)

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/plugins/flow.js]: Yeah, that are the most hard to debug test failures :man_facepalming:

**Pull request merged**